Story Arcs Basic Architecture

AaronGullickson opened the pull request:
This PR adds a major new feature, Story Arcs. I won't go over what story arcs are here as that is covered extensively on the [Project Details Page](https://github.com/orgs/MegaMek/projects/7?pane=info) for the GH project. Here I will discuss relevant information for reviewers.

This PR adds the underlying architecture for Story Arcs, and provides a test story arc named "Young Wolves" that players can use to try out the concept. It does not contain any ability to create story arcs apart from writing literal XML code. That is the next major planned PR in the project. 

I have tried to design the architecture of story arcs so that additional features are relatively easy to implement by simply extending the abstract classes of `StoryPoint` and `StoryTrigger` to new classes that can do particular things. In particular, as we develop story arcs further, we will likely see lots of requests to expand the use of `StoryTriggers` which can be used to make changes to the underlying campaign based on a certain result from a `StoryPoint`. Currently, I only built the triggers necessary for the example story arc, and one could imagine all kinds of other things that could be added. `StoryPoints` provide much more coverage of the kinds of things that are needed, but I can see the need to expand the number of "Check" StoryPoints which check a condition of something in the campaign.  

Although this PR changes/adds a lot of files, most of the work is contained within story arcs and should not affect base code. Any additional new features that I needed to get story arcs up and off the ground that involved significant changes to the base MHQ code have already been done as separate PRs and merged into master. Therefore, this code should touch existing code relatively lightly and any errors remaining will likely just mess up story arcs, not MHQ more broadly.

Given the extensiveness of the code, I am holding off on writing tests for the moment as that would delay this PR even further. On the project page, I do list adding tests for story arcs as one of the future goals, but it is prioritized lower than getting out an editor.

Duuvian commented:
This sounds cool. If I may make a suggestion, having a static pool of units for the opponent, allies, neutral/civ and neutral aggressive teams if present in the story would be awesome.  I think I saw something about static forces for stratcon in the pulse in the past few months but I can't remember what it was about.

Another suggestion would be to add a way to import unit lists to populate a static pool. That way canon unit lists could be brought in as an opponent. If it has a fluff section that can optionally list canon historical deployment areas it would also help with a petty issue I have with contracts in that I don't know what would be a reasonably canon unit I could name the bot opponent. Being able to import a unit with it's unitlist within it's inception and disbandment dates would also be useful for that, especially if it's capable of importing partial units like a company or regiment out of a multi regiment unit. It's not terribly imporant and seems like a complicated suggestion on my part though.

AaronGullickson commented:
Thanks for your thoughts, @Duuvian.  By static, I was referring to the ability for the story arc creator to specify a specific set of OpFor units, rather than using one of our random force generating methods. It has the upside of creating more realistic and balanced OpFors, but has the downside of not necessarily being balanced against the player's forces (although I can imagine ways to enforce this by limiting the units a player can deploy). I also imagine the possibility of mixing the two types in cases where the scenario requires a specific subset of units (say a commanding officer). 

I think what you are talking about is the ability to track the overall OpFor forces across multiple scenarios, so you can account for unit losses and the like. I agree that this would be a nice feature, but I doubt it will be part of this PR. The goals here are more basic, to give us an architecture for this kind of play that we can build upon.

Now on the idea of canon unit lists, we have talked about the possibility of including this kind of information (specifically, what units are garrisoning what planets) in our universe data. That wouldn't be part of this PR but it is something that if implemented, the story arcs (and other things like AtB/stratcon) could make use of. There is no feasible way that anyone is going to generate full TO&Es that are dynamic over time, but we could set up some basic parameters like average skill rating, tech level, percent clan tech, etc. that can help inform a random force generator that can then produce more realistic units. So you could say something like "give me a heavy mech lance from the 9th Marik Militia" and you get something that makes sense for the canon unit and the time period. But that is a very different feature than what I am doing here.

AaronGullickson commented:
Ok, I think its time for an update. I have made a lot of progress. The following changes have been made:

1. I have added a StoryNarrativeEvent which pops up a dialog with narrative information for the player.
2. I have added a StoryChoiceEvent which pops up a dialog which asks the user to make a choice which can then affect the StoryOutcome.
3. I have added StoryPortrait and StorySplash objects which extend AbstractIcon and contain story related portraits (used for Personalities below) and splash images. These can both be added to popup dialogs to enhance the narrative immersion. 
3. I have made it possible with some DirectoryItem-fu for creators to keep a data subdirectory in their story arc and have story arc specific portrait and splash images loaded from there.  
4. I have added a Personality object that can be associated with any StoryEvent. The main purpose of this object is that it has a portrait which is applied to any popup dialog from that StoryEvent. This can be used to make StoryNarrativeEvents and StoryChoiceEvents appear like they are being delivered by a person. Given that this works well, I will not be creating a separate ShoutStoryEvent. I will also probably push back the full rpg style dialog to a future enhancement as the basic functionality of this already provided with the StoryChoiceEvent tied to a Personality. 
5. I have done some more work on the "Young Wolves" story arc that I am building to test out the full capabilities of story arcs. Natasha Kerensky is in the game now, y'all.

What I still envision doing before the PR is ready:

1. A TravelStoryEvent as described above. 
2. I am considering adding a CreateCharacterEvent.java that will give players an XP pool they can use to create their own ego character for the story arc. Typically this would be the first event in a campaign if it allows for it, although I could imagine other uses.
3. Make scenarios better. This is the big one described in the original post. It might actually make more sense to do this in a separate PR as my idea is that the improvements to scenarios would not be specific to Story Arcs, but if I did it that way, I would probably not release this PR until that one was ready. 
4. Finish up my demo story arc. 
4. General code cleanup. I have been trying to document well and follow norms as I go, but I know that some touchup is necessary (as the nagging checks here on GH are also reminding me). 

One nice thing about this PR is that it has a fairly light touch with existing code. Mostly I had to add some code to get story arcs in and loaded, but otherwise even if story arcs are full of bugs, it has a low chance of affecting other areas of MHQ.

AaronGullickson commented:
One other thing I forgot in the previous list:

* A GUI element on the Command Center tab that lists current story arc objectives from all active StoryEvents

Windchild292 commented on MekHQ/src/mekhq/gui/dialog/DataLoadingDialog.java:
```diff
@@ -168,6 +177,20 @@ public Campaign doInBackground() throws Exception {
 
             Campaign campaign;
+
+            if(useStoryArc) {
```
spacing

Windchild292 commented on MekHQ/src/mekhq/gui/dialog/DataLoadingDialog.java:
```diff
@@ -314,6 +338,12 @@ public void done() {
                 frame.setVisible(false);
                 frame.dispose();
                 app.showNewView();
+                if(null != storyArcStub) {
+                    StoryArc storyArc = storyArcStub.loadStoryArc(campaign);
+                    if(null != storyArc) {
```
Also here

HammerGS commented:
![image](https://github.com/MegaMek/mekhq/assets/13752452/db44202d-66c9-44ff-ab52-4a4398e06d5e)

AaronGullickson commented:
![fb2b58d7-d7bc-44be-a154-b400f460d4da_text](https://github.com/MegaMek/mekhq/assets/16107446/928936f9-2831-45a4-806c-87a96d694ed0)

HammerGS commented:
> ![fb2b58d7-d7bc-44be-a154-b400f460d4da_text](https://private-user-images.githubusercontent.com/16107446/313432096-928936f9-2831-45a4-806c-87a96d694ed0.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA2MjUzNjQsIm5iZiI6MTcxMDYyNTA2NCwicGF0aCI6Ii8xNjEwNzQ0Ni8zMTM0MzIwOTYtOTI4OTM2ZjktMjgzMS00NWE0LTgwNmMtODdhOTZkNjk0ZWQwLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzE2VDIxMzc0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY2NGI0NzE5OWIzYzQ4ODE3MmJlNjNhMjU1NzRhYWM2MDcxZDU0ODgwYTUyYjZkODU1NTZmZjQ4NjNlZGE0ZmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.xcO8DO7e4Twk0aDxwgWOaiLk0p_6rQhcExT4bc9IXvQ) [ ![fb2b58d7-d7bc-44be-a154-b400f460d4da_text](https://private-user-images.githubusercontent.com/16107446/313432096-928936f9-2831-45a4-806c-87a96d694ed0.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA2MjUzNjQsIm5iZiI6MTcxMDYyNTA2NCwicGF0aCI6Ii8xNjEwNzQ0Ni8zMTM0MzIwOTYtOTI4OTM2ZjktMjgzMS00NWE0LTgwNmMtODdhOTZkNjk0ZWQwLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzE2VDIxMzc0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY2NGI0NzE5OWIzYzQ4ODE3MmJlNjNhMjU1NzRhYWM2MDcxZDU0ODgwYTUyYjZkODU1NTZmZjQ4NjNlZGE0ZmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.xcO8DO7e4Twk0aDxwgWOaiLk0p_6rQhcExT4bc9IXvQ) ](https://private-user-images.githubusercontent.com/16107446/313432096-928936f9-2831-45a4-806c-87a96d694ed0.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA2MjUzNjQsIm5iZiI6MTcxMDYyNTA2NCwicGF0aCI6Ii8xNjEwNzQ0Ni8zMTM0MzIwOTYtOTI4OTM2ZjktMjgzMS00NWE0LTgwNmMtODdhOTZkNjk0ZWQwLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzE2VDIxMzc0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY2NGI0NzE5OWIzYzQ4ODE3MmJlNjNhMjU1NzRhYWM2MDcxZDU0ODgwYTUyYjZkODU1NTZmZjQ4NjNlZGE0ZmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.xcO8DO7e4Twk0aDxwgWOaiLk0p_6rQhcExT4bc9IXvQ) [ ](https://private-user-images.githubusercontent.com/16107446/313432096-928936f9-2831-45a4-806c-87a96d694ed0.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA2MjUzNjQsIm5iZiI6MTcxMDYyNTA2NCwicGF0aCI6Ii8xNjEwNzQ0Ni8zMTM0MzIwOTYtOTI4OTM2ZjktMjgzMS00NWE0LTgwNmMtODdhOTZkNjk0ZWQwLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzE2VDIxMzc0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY2NGI0NzE5OWIzYzQ4ODE3MmJlNjNhMjU1NzRhYWM2MDcxZDU0ODgwYTUyYjZkODU1NTZmZjQ4NjNlZGE0ZmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.xcO8DO7e4Twk0aDxwgWOaiLk0p_6rQhcExT4bc9IXvQ)

Good to see. Hit me up on Discord and I'll catch you up on things when you have a few.

Sleet01 commented on MekHQ/src/mekhq/MHQStaticDirectoryManager.java:
```diff
@@ -93,6 +100,67 @@ private static void initializeAwardIcons() {
+        // Read in and parse MekHQ's force icon folder only when first called or when refreshed
+        if (parseStorySplashDirectory) {
+            // Set parseForceIconDirectory to false to avoid parsing repeatedly when something fails
+            parseStorySplashDirectory = false;
```
I'm not seeing where this boolean would get reset so that a refresh call would re-load the directory items.

Sleet01 replied:
NM, just lower down.

Sleet01 commented on MekHQ/src/mekhq/campaign/Campaign.java:
```diff
@@ -4084,6 +4105,29 @@ public void setCampaignOptions(CampaignOptions options) {
         campaignOptions = options;
     }
 
+    public StoryArc getStoryArc() { return storyArc; }
```
I believe the style guide requires this to be 3 lines: https://github.com/MegaMek/megamek/wiki/MegaMek-Coding-Style-Guide#braces

Sleet01 commented on MekHQ/src/mekhq/campaign/storyarc/StoryArc.java:
```diff
@@ -0,0 +1,521 @@
+        customStringVariables = new LinkedHashMap<>();
+    }
+
+    public void setCampaign(Campaign c) { this.campaign = c; }
+
+    public Campaign getCampaign() { return campaign; }
+
+    private void setTitle(String t) { this.title = t; }
+
+    public String getTitle() { return this.title; }
+
+    public String getDetails() { return details; }
+
+    private void setDetails(String d) { this.details = d; }
+
+    public String getDescription() { return this.description; }
+
+    private void setDescription(String d) { this.description = d; }
+
+    private void setStoryLoadingType(StoryLoadingType type) { this.storyLoadingType = type; }
+
+    public StoryLoadingType getStoryLoadingType() { return storyLoadingType; }
+
+    private void setStartingPointId(UUID u) { this.startingPointId = u; }
+
+    private UUID getStartingPointId() { return startingPointId; }
+
+    public void setInitCampaignPath(String s) { this.initCampaignPath =s; }
```
These all need to be three-line declarations with the closing brace on its own line.

AaronGullickson replied:
@Sleet01 Thanks for that catch. I will go through the PR and fix any of those I see.

AaronGullickson replied:
Resolved in latest. I looked through all the files and changed all instances of one line setter/getters to three lines.

Sleet01 commented on MekHQ/src/mekhq/gui/dialog/StoryDialog.java:
```diff
@@ -0,0 +1,123 @@
+        JPanel buttonPanel = new JPanel(new BorderLayout());
+
+        doneButton = new JButton("Done");
+        //doneButton.setMnemonic('o');
```
Should probably be removed.

Sleet01 reviewed (approved):
Looks really good!
